Makefile

```diff
@@ -0,0 +1,65 @@
+# Copyright (C) 2019 Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+CRC_VERSION = 0.0.1
+COMMIT_SHA=$(shell git rev-parse --short HEAD)
+
+# Go and compilation related variables
+BUILD_DIR ?= out
+
+GOOS ?= $(shell go env GOOS)
+GOARCH ?= $(shell go env GOARCH)
+ORG := github.com/code-ready
+REPOPATH ?= $(ORG)/crc
+ifeq ($(GOOS),windows)
+	IS_EXE := .exe
+endif
+CRC_BINARY ?= $(GOPATH)/bin/crc$(IS_EXE)
+TIMEOUT ?= 10800s
+PACKAGES := go list ./... | grep -v /out
+SOURCE_DIRS = cmd pkg
+
+
+# Linker flags
+VERSION_VARIABLES := -X $(REPOPATH)/pkg/version.crcVersion=$(CRC_VERSION) \
+	-X $(REPOPATH)/pkg/version.commitSha=$(COMMIT_SHA)
+
+LDFLAGS := $(VERSION_VARIABLES)
+
+# Start of the actual build targets
+
+.PHONY: $(GOPATH)/bin/crc$(IS_EXE)
+$(GOPATH)/bin/crc$(IS_EXE): $(ADDON_ASSET_FILE) ## Builds the binary into $GOPATH/bin
+	go install -tags "$(BUILD_TAGS_SYSTEMTRAY)" -pkgdir=$(ADDON_BINDATA_DIR) -ldflags="$(VERSION_VARIABLES)" ./cmd/crc
+
+$(BUILD_DIR)/$(GOOS)-$(GOARCH):
+	mkdir -p $(BUILD_DIR)/$(GOOS)-$(GOARCH)
+
+$(BUILD_DIR)/darwin-amd64/crc: $(BUILD_DIR)/$(GOOS)-$(GOARCH) ## Cross compiles the darwin executable and places it in $(BUILD_DIR)/darwin-amd64/crc
+	CGO_ENABLED=0 GOARCH=amd64 GOOS=darwin go build --installsuffix cgo -ldflags="$(LDFLAGS)" -o $(BUILD_DIR)/darwin-amd64/crc ./cmd/crc
+
+$(BUILD_DIR)/linux-amd64/crc: $(ADDON_ASSET_FILE) $(BUILD_DIR)/$(GOOS)-$(GOARCH) ## Cross compiles the linux executable and places it in $(BUILD_DIR)/linux-amd64/crc
+	CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build --installsuffix cgo -ldflags="$(LDFLAGS)" -o $(BUILD_DIR)/linux-amd64/crc ./cmd/crc
+
+$(BUILD_DIR)/windows-amd64/crc.exe: $(ADDON_ASSET_FILE) $(BUILD_DIR)/$(GOOS)-$(GOARCH) ## Cross compiles the windows executable and places it in $(BUILD_DIR)/windows-amd64/crc
+	CGO_ENABLED=0 GOARCH=amd64 GOOS=windows go build --installsuffix cgo -ldflags="$(LDFLAGS)" -o $(BUILD_DIR)/windows-amd64/crc.exe ./cmd/crc
```

Comment:
The Windows cross build is failing now as it has an external dependency, need to fix that 
```
$ make cross
mkdir -p out/linux-amd64
CGO_ENABLED=0 GOARCH=amd64 GOOS=darwin go build --installsuffix cgo -ldflags="-X github.com/code-ready/crc/pkg/version.crcVersion=0.0.1 -X github.com/code-ready/crc/pkg/version.commitSha=459ba34" -o out/darwin-amd64/crc ./cmd/crc
CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build --installsuffix cgo -ldflags="-X github.com/code-ready/crc/pkg/version.crcVersion=0.0.1 -X github.com/code-ready/crc/pkg/version.commitSha=459ba34" -o out/linux-amd64/crc ./cmd/crc
CGO_ENABLED=0 GOARCH=amd64 GOOS=windows go build --installsuffix cgo -ldflags="-X github.com/code-ready/crc/pkg/version.crcVersion=0.0.1 -X github.com/code-ready/crc/pkg/version.commitSha=459ba34" -o out/windows-amd64/crc.exe ./cmd/crc
../../spf13/cobra/command_win.go:9:2: cannot find package "github.com/inconshreveable/mousetrap" in any of:
	/usr/lib/golang/src/github.com/inconshreveable/mousetrap (from $GOROOT)
	/home/lmohanty/go/src/github.com/inconshreveable/mousetrap (from $GOPATH)
make: *** [Makefile:56: out/windows-amd64/crc.exe] Error 1
```

Reply:
Also I am not sure if we need `CGO_ENABLED=0` and `--installsuffix cgo`, looking in to that

Reply:
Ok, we definitly need ` CGO_ENABLED=0` , refer https://dave.cheney.net/tag/cgo

Reply:
As per this article `By default cgo is disabled when cross compiling. Normally this isn’t a problem if your project is pure Go. When you mix in dependencies on C libraries, you either have to give up the option to cross compile your product, or you have to invest time in finding and maintaining cross compilation C toolchains for all your targets.`  CGO is by default disabled so why we need to put and env to do same thing?

Reply:
@praveenkumar `explict is better than implicit` when you add dependencies which might need CGO how do you make sure the default does not change unless you are explicitly using `CGO_ENABLED=0`?

Reply:
@LalatenduMohanty I agree about `explict is better than implicit` but if you add dependencies which might need CGO then your cross-compile automatic fails and you need to add it explicit using `CGO_ENABLED=1` , default behaviour is there for user not to repeat those.

Reply:
> if you add dependencies which might need CGO then your cross-compile automatic fails and you need to add it explicit using `CGO_ENABLED=1` , default behaviour is there for user not to repeat those.

Using `CGO_ENABLED=1` is a big deal in golang as you can find it in various articles. It increases teh compilation time etc. So  explicitly mentioning it `CGO_ENABLED=1` would something I would prefer. Because when someone will make it `CGO_ENABLED=1` it would be tracked in the code change (we will cross check and careful about it in PR review) and commits which is good thing IMO @gbraad @zeenix @anjannath WDYT?

Reply:
I rather have less cluttered makefiles... so default should NOT be set again or explicitly; it does not help with readability

Reply:
@praveenkumar @gbraad I am trying to validate that `CGO` is disabled by default in recent go versions. However I am yet to find any articles to prove this. If you can point me to any articles it would great.

Reply:
Ok, got the reference https://github.com/golang/go/issues/9344#issuecomment-69944514